typing/typecore.ml

```diff
@@ -4184,10 +4210,15 @@ and type_argument ?explanation ?recarg env mode sarg ty_expected' ty_expected =
       end;
       let rec make_args args ty_fun =
         match (expand_head env ty_fun).desc with
-        | Tarrow (l,ty_arg,ty_fun,_) when is_optional l ->
-            let ty = option_none env (instance ty_arg) sarg.pexp_loc in
+        | Tarrow ((l,marg,mret),ty_arg,ty_fun,_) when is_optional l ->
+            begin match Types.Alloc_mode.submode mret Alloc_heap with
```

Comment:
I'm not convinced! Optional arguments are not necessarily eliminated by the very next argument, because you can write:
```ocaml
let f ?x ~y () = ...
```
The type of `f ()` is `y:_ -> _` and the return mode of the optional argument affects the mode of this closure.

Reply:
Never mind, I thought you were referring to type_application here, not type_argument. It's probably fine to remove the condition here.